src/dust/dust.cpp

```diff
@@ -227,6 +285,81 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
   return dust;
 }
 
+//----------------------------------------------------------------------------------------
+//! \fn  TaskStatus Dust::UpdateDustStoppingTime
+//  \brief Wrapper function for update dust stopping time
+template <Coordinates GEOM>
+TaskStatus UpdateDustStoppingTime(MeshData<Real> *md) {
+
+  using parthenon::MakePackDescriptor;
+
+  auto pm = md->GetParentPointer();
+
+  auto &dust_pkg = pm->packages.Get("dust");
+  if ((!dust_pkg->template Param<bool>("enable_coagulation"))) {
+    return TaskStatus::complete;
+  }
+
+  auto &resolved_pkgs = pm->resolved_packages;
+
+  IndexRange ib = md->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = md->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = md->GetBoundsK(IndexDomain::interior);
+
+  const int nspecies = dust_pkg->template Param<int>("nspecies");
+
+  const bool isurf_den = cgsunit->isurface_den;
+  ParArray1D<Real> dust_size = dust_pkg->template Param<ParArray1D<Real>>("sizes");
+  static auto desc =
+      MakePackDescriptor<gas::prim::density, gas::prim::pressure, dust::stopping_time>(
+          resolved_pkgs.get());
+
+  auto vmesh = desc.GetPack(md);
+
+  auto &gas_pkg = pm->packages.Get("gas");
+
+  const Real gamma = gas_pkg->template Param<Real>("adiabatic_index");
+  const Real rho_p = dust_pkg->template Param<Real>("rho_p");
+
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "Dust::DustStoppingTime2", parthenon::DevExecSpace(), 0,
+      md->NumBlocks() - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+        const Real dens_g = vmesh(b, gas::prim::density(0), k, j, i);
+        // usually the Stokes number depends gas density, unless constant is used
+        // stopping_time = Stokes_number/Omega_k
+
+        if (isurf_den) {
+          geometry::Coords<GEOM> coords(vmesh.GetCoordinates(b), k, j, i);
+          // for surface denstiy, Stokes number = Pi/2*rho_p*s_p/sigma_g
+          // calculate the Keplerian Omega at mid-plane
+          Real rad;
+          if constexpr (GEOM == Coordinates::cylindrical) {
+            rad = coords.hx2v(); // cylindrical-rad
+          } else {
+            rad = coords.hx3v();
+          }
+          Real Omega_k = 1.0 / std::sqrt(rad) / rad;
+          for (int n = 0; n < nspecies; ++n) {
+            const Real St = rho_p * dust_size(n) / dens_g;
+            vmesh(b, dust::stopping_time(n), k, j, i) = St / Omega_k;
+          }
+
+        } else {
+
+          // for density (g/cc), Stokes number = sqrt(Pi/8)*rho_p*s_p*Omega_k/rho_g/Cs
+          const Real pres = vmesh(b, gas::prim::pressure(0), k, j, i);
+          const Real cs = std::sqrt(gamma * pres / dens_g);
```

Comment:
Whenever you need `cs`, you should get it from the eos via the bulk modulus, e.g., `BulkModulusFromDensityInternalEnergy`.